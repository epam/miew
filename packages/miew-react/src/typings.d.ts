/**
 * Default CSS definition for typescript,
 * will be overridden with file-specific definitions by rollup
 */
declare module '*.scss' {
  const content: { [className: string]: string }
  export default content
}

type SvgrComponent = React.StatelessComponent<React.SVGAttributes<SVGElement>>

declare module '*.svg' {
  // eslint-disable-next-line no-unused-vars
  const svgUrl: string
  const svgComponent: SvgrComponent
  export default svgComponent
}
