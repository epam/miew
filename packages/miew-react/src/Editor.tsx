import useResizeObserver from 'use-resize-observer'
import { AppContainer } from './components/App'

const MEDIA_SIZES = {
  smallWidth: 800,
  smallHeight: 400
}

const Editor = () => {
  const { ref, width, height } = useResizeObserver<HTMLDivElement>()

  const isSizeSmall =
    (height && height <= MEDIA_SIZES.smallHeight) ||
    (width && width <= MEDIA_SIZES.smallWidth)

  const containerStyle = {
    height: '100%',
    width: '100%',
    backgroundColor: isSizeSmall ? '#eee' : '#fff'
  }

  return (
    <div css={containerStyle} ref={ref}>
      <AppContainer />
    </div>
  )
}

export { Editor }
