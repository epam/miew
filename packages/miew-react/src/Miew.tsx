import useResizeObserver from 'use-resize-observer'
import clsx from 'clsx'
import classes from './Miew.module.scss'

const MEDIA_SIZES = {
  smallWidth: 800,
  smallHeight: 400
}

type Props = {
  text: string
}

function Miew({ text }: Props) {
  const { ref, width, height } = useResizeObserver<HTMLDivElement>()

  return (
    <div
      className={clsx(classes.miew, {
        [classes.small]:
          (height && height <= MEDIA_SIZES.smallHeight) ||
          (width && width <= MEDIA_SIZES.smallWidth)
      })}
      ref={ref}
    >
      <p>HELLO, I am Miew!</p>
      <p>{text}</p>
      <p>
        Width: {width} Height: {height}
      </p>
    </div>
  )
}

export { Miew }
