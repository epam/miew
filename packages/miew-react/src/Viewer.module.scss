.viewer {
  background-color: olive;
  color: #fff;
  width: 100%;
  height: 100%;
}

.small {
  background-color: royalblue;
}