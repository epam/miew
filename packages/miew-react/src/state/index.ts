export * from './store'
export * from './rootSaga'
export * from './hooks'
