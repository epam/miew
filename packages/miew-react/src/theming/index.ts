export * from './defaultTheme'
export * from './theme.types'
