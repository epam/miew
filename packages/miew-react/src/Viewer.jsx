import Miew from 'miew';
import PropTypes from 'prop-types';
import React, { useEffect, useLayoutEffect, useRef } from 'react';
import * as styles from './Viewer.module.scss';
import 'miew/dist/Miew.css';

/**
 * Creates and initializes a Miew instance
 * @param {Object} miewRef - Ref to store the Miew instance
 * @param {Object} options - Miew initialization options
 * @param {Function} onInit - Callback function called after successful initialization
 * @param {Function} onError - Callback function called when an error occurs
 */
function createMiewRef(miewRef, options, onInit, onError) {
  if (!miewRef.current) {
    try {
      const miew = new Miew(options);
      if (miew.init()) {
        miewRef.current = miew;
        miew.run();

        if (typeof onInit === 'function') {
          onInit(miew);
        }
      } else {
        const error = new Error('Failed to initialize Miew');
        if (typeof onError === 'function') {
          onError(error);
        }
      }
    } catch (error) {
      if (typeof onError === 'function') {
        onError(error);
      }
    }
  }
}

/**
 * Properly destroys a Miew instance and cleans up resources
 * @param {Object} miewRef - Ref containing the Miew instance
 */
function destroyMiewRef(miewRef) {
  if (miewRef.current) {
    miewRef.current.term();
  }
  miewRef.current = null;
}

/**
 * React component wrapper for Miew 3D molecular viewer
 * @param {Object} props - Component props
 * @param {Function} props.onInit - Callback function called after successful Miew initialization
 * @param {Function} props.onError - Callback function called when an error occurs during initialization
 * @param {Object} props.options - Miew configuration options
 * @param {*} props.theme - Legacy prop kept for backwards compatibility, will be removed in future versions
 * @returns {JSX.Element} The Viewer component
 */
export default function Viewer({ onInit, onError, options, theme }) {
  const miewRef = useRef();
  const rootRef = useRef();

  // Deprecation warning for theme prop
  useEffect(() => {
    if (theme !== undefined) {
      console.warn(
        'Warning: The "theme" prop is deprecated and will be removed in future versions of miew-react. It is kept for backwards compatibility only.',
      );
    }
  }, [theme]);

  useLayoutEffect(() => {
    const settings = { axes: false, fps: false, ...options?.settings };
    createMiewRef(miewRef, { ...options, container: rootRef.current, settings }, onInit, onError);
    return () => destroyMiewRef(miewRef);
  }, [onInit, onError, options]);

  return (
    <div className={styles.root} ref={rootRef}>
      Viewer
    </div>
  );
}

Viewer.propTypes = {
  onInit: PropTypes.func,
  onError: PropTypes.func,
  options: PropTypes.shape({
    load: PropTypes.string, // Molecule identifier or file path
    type: PropTypes.string, // File type (pdb, cif, etc.)
    settings: PropTypes.object, // Viewer settings (axes, fps, autoRotation, bg, etc.)
    settingsCookie: PropTypes.string, // Cookie name for storing settings
    cookiePath: PropTypes.string, // Cookie path
    reps: PropTypes.arrayOf(PropTypes.object), // Representations array
    container: PropTypes.instanceOf(typeof HTMLElement !== 'undefined' ? HTMLElement : Object), // DOM container element
  }),
  /** @deprecated This prop is deprecated and will be removed in future versions */
  theme: PropTypes.any,
};

Viewer.defaultProps = {
  onInit: undefined,
  onError: undefined,
  options: undefined,
  theme: undefined,
};
