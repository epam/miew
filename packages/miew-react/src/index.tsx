import { Miew } from './Miew'

export default Miew
