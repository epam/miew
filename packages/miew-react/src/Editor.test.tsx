import { render } from '@testing-library/react'
import { Editor } from 'src/Editor'

// TODO create mock folder, and move all mocks there
jest.mock('use-resize-observer', () =>
  jest.fn().mockReturnValue({ ref: null, width: 100, height: 100 })
)

describe('Editor component', () => {
  it('should be rendered', () => {
    expect(render(<Editor />)).toMatchSnapshot()
  })
})
