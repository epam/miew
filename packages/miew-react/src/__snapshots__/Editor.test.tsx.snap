// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor component should be rendered 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .emotion-0 {
  height: 100%;
  width: 100%;
  background-color: #eee;
}

.emotion-1 {
  color: #383838;
  font-size: 14px;
}

.emotion-2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 500;
  font-size: 0.875rem;
  line-height: 1.75;
  letter-spacing: 0.02857em;
  text-transform: uppercase;
  min-width: 64px;
  padding: 6px 8px;
  border-radius: 4px;
  -webkit-transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  color: #1976d2;
  border: 1px solid #333;
  margin: 10px;
}

.emotion-2::-moz-focus-inner {
  border-style: none;
}

.emotion-2.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-2 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-2:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: rgba(25, 118, 210, 0.04);
}

@media (hover: none) {
  .emotion-2:hover {
    background-color: transparent;
  }
}

.emotion-2.Mui-disabled {
  color: rgba(0, 0, 0, 0.26);
}

.emotion-3 {
  overflow: hidden;
  pointer-events: none;
  position: absolute;
  z-index: 0;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  border-radius: inherit;
}

.emotion-4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 500;
  font-size: 0.875rem;
  line-height: 1.75;
  letter-spacing: 0.02857em;
  text-transform: uppercase;
  min-width: 64px;
  padding: 6px 8px;
  border-radius: 4px;
  -webkit-transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  color: #9c27b0;
  border: 1px solid #333;
  margin: 10px;
}

.emotion-4::-moz-focus-inner {
  border-style: none;
}

.emotion-4.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-4 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-4:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: rgba(156, 39, 176, 0.04);
}

@media (hover: none) {
  .emotion-4:hover {
    background-color: transparent;
  }
}

.emotion-4.Mui-disabled {
  color: rgba(0, 0, 0, 0.26);
}

<body>
    <div>
      <div
        class="emotion-0"
      >
        <div
          class="emotion-1"
        >
          <p>
            HELLO, I am Miew!
          </p>
          <p>
            App is not initialized
          </p>
          <button
            class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root emotion-2"
            tabindex="0"
            type="button"
          >
            Re-initialize app
            <span
              class="MuiTouchRipple-root emotion-3"
            />
          </button>
          <button
            class="MuiButton-root MuiButton-text MuiButton-textSecondary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root emotion-4"
            tabindex="0"
            type="button"
          >
            Simulate error
            <span
              class="MuiTouchRipple-root emotion-3"
            />
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": .emotion-0 {
  height: 100%;
  width: 100%;
  background-color: #eee;
}

.emotion-1 {
  color: #383838;
  font-size: 14px;
}

.emotion-2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 500;
  font-size: 0.875rem;
  line-height: 1.75;
  letter-spacing: 0.02857em;
  text-transform: uppercase;
  min-width: 64px;
  padding: 6px 8px;
  border-radius: 4px;
  -webkit-transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  color: #1976d2;
  border: 1px solid #333;
  margin: 10px;
}

.emotion-2::-moz-focus-inner {
  border-style: none;
}

.emotion-2.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-2 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-2:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: rgba(25, 118, 210, 0.04);
}

@media (hover: none) {
  .emotion-2:hover {
    background-color: transparent;
  }
}

.emotion-2.Mui-disabled {
  color: rgba(0, 0, 0, 0.26);
}

.emotion-3 {
  overflow: hidden;
  pointer-events: none;
  position: absolute;
  z-index: 0;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  border-radius: inherit;
}

.emotion-4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 500;
  font-size: 0.875rem;
  line-height: 1.75;
  letter-spacing: 0.02857em;
  text-transform: uppercase;
  min-width: 64px;
  padding: 6px 8px;
  border-radius: 4px;
  -webkit-transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  color: #9c27b0;
  border: 1px solid #333;
  margin: 10px;
}

.emotion-4::-moz-focus-inner {
  border-style: none;
}

.emotion-4.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-4 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-4:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: rgba(156, 39, 176, 0.04);
}

@media (hover: none) {
  .emotion-4:hover {
    background-color: transparent;
  }
}

.emotion-4.Mui-disabled {
  color: rgba(0, 0, 0, 0.26);
}

<div>
    <div
      class="emotion-0"
    >
      <div
        class="emotion-1"
      >
        <p>
          HELLO, I am Miew!
        </p>
        <p>
          App is not initialized
        </p>
        <button
          class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root emotion-2"
          tabindex="0"
          type="button"
        >
          Re-initialize app
          <span
            class="MuiTouchRipple-root emotion-3"
          />
        </button>
        <button
          class="MuiButton-root MuiButton-text MuiButton-textSecondary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root emotion-4"
          tabindex="0"
          type="button"
        >
          Simulate error
          <span
            class="MuiTouchRipple-root emotion-3"
          />
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
